### DISCUSSION TITLE
How to run a slow task on button:click?

### DISCUSSION THREAD

**Discussion by danechitoaie** [leptos-rs/leptos#1848] [Q&A] [answered]
I'm trying to PGP encrypt a string client side when user clicks on a button and this seems to be super slow.
First when running in debug/dev mode with `trunk serve --open` it takes ages for the page to load with the code bellow.
If I compile in production mode `trunk build --release` then it's faster but still loads much slower (~3-5s) compared to a JavaScript equivalent.

1. How can I run a task take takes longer time to process in Leptos as code bellow freezes the UI for few seconds.
2. Any idea how I can make it run faster? I would like it to be similar to JS version in regards to speed.

```rust
use base64::{engine::general_purpose, Engine as _};
use leptos::*;
use sequoia_openpgp::serialize::stream::{Encryptor, LiteralWriter, Message};
use std::io::Write;

#[component]
pub fn App() -> impl IntoView {
    let (data, set_data) = create_signal("".to_string());

    view! {
        <div>
            <pre>{move || data.get()}</pre>
            <button on:click=move |_| {
                    let mut sink = vec![];
                    let message = Encryptor::with_passwords(Message::new(&mut sink), Some("password"))
                        .build()
                        .unwrap();
                    let mut w = LiteralWriter::new(message).build().unwrap();
                    let _ = w.write_all(b"Hello world.");
                    let _ = w.finalize();
                    let b64 = general_purpose::STANDARD.encode(sink);

                    set_data.set(b64.clone());
                }
            >
                "Encrypt"
            </button>
        </div>
    }
}

pub fn main() {
    mount_to_body(|| {
        view! {
            <App />
        }
    })
}
```

**Comment by gbj** [comment]
I'm having trouble getting `sequoia-openpgp` even to compile, unfortunately, or I'd be interested to explore more. 

I don't think the question really has anything to do with Leptos — it's blocking on the processing time inside your callback, right? I'm totally unfamiliar with the domain here unfortunately.

What's your equivalent JS code? If it's using some browser built-ins you might as well use those instead of compiling in additional Rust dependencies.

**Reply by danechitoaie** [reply]
What errors are you getting with `sequoia-openpgp`?
Basically what I did for the above example was:

```
cargo add sequoia-openpgp --no-default-features --features=crypto-rust,allow-experimental-crypto,allow-variable-time-crypto
```

and then I was getting this error:
```
error[E0599]: no function or associated item named `now` found for struct `Utc` in the current scope
```

and then I've added `chrono` with `cargo add` and after that I was able to compile.
Felt strange to me that if sequoia requires chrono why isn't it a dependency of sequoia that gets installed automatically and I had to manually add it, but anyway I'm new to Rust so maybe I'm missing something.


> I don't think the question really has anything to do with Leptos — it's blocking on the processing time inside your callback, right? I'm totally unfamiliar with the domain here unfortunately.

Yes, I think so. Maybe it's more of a wasm question in general? Basically the problem is how do you run some code that blocks in a way that doesn't freeze the UI? Can I run an async function (that will have equivalent async code to encrypt instead of the sync blocking one)? Or should it be offloading it in a webworker? Can that be done from wasm?


> What's your equivalent JS code? If it's using some browser built-ins you might as well use those instead of compiling in additional Rust dependencies.
The code I'm trying to port is using https://www.npmjs.com/package/openpgp 
```javascript
const message = await openpgp.createMessage({ binary: new Uint8Array([0x01, 0x01, 0x01]) });
    const encrypted = await openpgp.encrypt({
        message, // input as Message object
        passwords: ['secret stuff'], // multiple passwords possible
    });
    console.log(encrypted);
```

**Reply by gbj** [reply]
Thanks, it was just a matter of the default features.

When I run this on my laptop in release mode I get a small lag after button click -- maybe a quarter of a second? So I suspect it might be a hardware difference with something CPU intensive.

In the browser, the only way to offload synchronous work in a non-blocking way that I'm aware of is to use a Web Worker, but I don't do much work like this. See [`leptos_worker`](https://docs.rs/leptos_workers/latest/leptos_workers/attr.worker.html) for one possible helper here.

Edit to add: of course this won't make it faster, but using a worker should at least make it non-blocking on slower hardware.

**Reply by danechitoaie** [reply]
I think the slowness is intended. I started looking at https://docs.rs/age/latest/age/ instead of sequoia and I've seen something mentioned that the slowness is due to trying to prevent timing attack thus making the encryption time constant and taking some constant time as well to generate the encryption key. I assume sequoia does the same since they use similar algorithms.

Anyway this https://docs.rs/age/latest/age/ seems to be having async methods https://docs.rs/age/latest/age/struct.Encryptor.html#method.wrap_async_output

Would I be able to execute an async method when a button is clicked? So even if it takes time at lease the browser doesn't freeze and I can show a loading message or something.

**Reply by gbj** [reply, marked as answer]
Oh, sure! Using async for an intentionally-slow task makes a ton of sense: an async approach here would basically do some work, set a timer or something, yield back to the browser until the timer goes off, then come back and do some more work, etc. Basically if what's causing the slowness is computation, async can't help you (although it can spread it out into non-blocking chunks). But if what's causing the slowness is intentionally blocking the thread for a period of time, async is ideal.

To an async task from the click handler just use [`spawn_local`](https://docs.rs/leptos/latest/leptos/fn.spawn_local.html).

**Reply by danechitoaie** [reply]
Thank you. I'll try to see if I can get it working.

**Reply by danechitoaie** [reply]
@gbj Quick question: Do you plan to add something like "leptos_worker" directly in Leptos so we have this feature OOTB? 
Yew seems to be having this ootb, and not as a 3rd party crate and it's more "trust worthy and convenient". I like Leptos the most out of all and I really like how "leptos_worker" is implemented as well, but it's worrisome that on its readme page it says "This crate is in an early alpha stage, use at your own risk. The API will change."

**Reply by gbj** [reply]
No, I don't plan on adding it into the core of the library. In general, whenever something can be implemented as a separate package without needing special access to framework internals, that is the best way for it to exist. In general, once something enters the Leptos repository itself it becomes ultimately my responsibility, and I can't take on the maintenance burden for every useful package.

**Reply by danechitoaie** [reply]
Ok, makes sense.
